cmd/sso.go

```diff
@@ -0,0 +1,259 @@
+		Short:   "Manage Single Sign-On (SSO) authentication for projects",
+	}
+
+	ssoProjectRef           string
+	ssoMetadataFile         string
+	ssoMetadataURL          string
+	ssoMetadata             bool
+	ssoAttributeMappingFile string
+	ssoDomains              []string
+	ssoAddDomains           []string
+	ssoRemoveDomains        []string
+	ssoOutput               = utils.OutputFlag(false /* allow env output */)
+
+	ssoAddCmd = &cobra.Command{
+		Use:     "add <type = saml> [flags]",
+		Short:   "Add a new SSO identity provider",
+		Args:    cobra.ExactArgs(1),
+		Example: `  supabase sso add saml --project-ref abcdefghijklmn --metadata-file ~/SAMLMetadata.xml`,
+		PreRun: func(cmd *cobra.Command, args []string) {
+			cobra.CheckErr(cmd.MarkFlagRequired("project-ref"))
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var params api.CreateProviderForProjectJSONRequestBody
+
+			switch strings.ToLower(args[0]) {
+			case "saml":
+				params.Type = api.Saml
+
+			default:
+				return errors.New("type must be saml")
+			}
+
+			fsys := afero.NewOsFs()
+
+			if ssoMetadataFile != "" {
+				data, err := readMetadataFile(fsys, ssoMetadataFile)
+				if err != nil {
+					return err
+				}
+
+				params.MetadataXml = &data
+			} else if ssoMetadataURL != "" {
+				// TODO: fetch and validate Metadata
+				params.MetadataUrl = &ssoMetadataURL
+			} else {
+				return errors.New("--metadata-file or --metadata-xml must be provided")
+			}
+
+			if ssoAttributeMappingFile != "" {
+				mapping, err := readAttributeMappingFile(fsys, ssoAttributeMappingFile)
+				if err != nil {
+					return err
+				}
+
+				params.AttributeMapping = mapping
+			}
+
+			params.Domains = &ssoDomains
+
+			return create.Run(cmd.Context(), ssoProjectRef, params, ssoOutput.Value)
+		},
+	}
+
+	ssoRemoveCmd = &cobra.Command{
+		Use:     "remove <provider-id> [flags]",
+		Short:   "Remove an existing SSO identity provider",
+		Args:    cobra.ExactArgs(1),
+		Example: `  supabase sso remove b5ae62f9-ef1d-4f11-a02b-731c8bbb11e8 --project-ref abcdefghijklmn`,
+		PreRun: func(cmd *cobra.Command, args []string) {
+			cobra.CheckErr(cmd.MarkFlagRequired("project-ref"))
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return remove.Run(cmd.Context(), ssoProjectRef, args[0], ssoOutput.Value)
+		},
+	}
+
+	ssoUpdateCmd = &cobra.Command{
+		Use:     "update <provider-id> [flags]",
+		Short:   "Update information about an SSO identity provider",
+		Args:    cobra.ExactArgs(1),
+		Example: `  supabase sso update b5ae62f9-ef1d-4f11-a02b-731c8bbb11e8 --project-ref abcdefghijklmn --add-domain example.com`,
+		PreRun: func(cmd *cobra.Command, args []string) {
+			cobra.CheckErr(cmd.MarkFlagRequired("project-ref"))
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var params api.UpdateProviderByIdJSONRequestBody
+
+			fsys := afero.NewOsFs()
+
+			if ssoMetadataFile != "" {
+				data, err := readMetadataFile(fsys, ssoMetadataFile)
```

Comment:
Can we move file reading into internals? I think it's fine to build the full `api.UpdateProviderByIdJSONRequestBody` object inside internals too.